install/install.go

```diff
@@ -0,0 +1,156 @@
+	rg, ok := cmdline["guid_root"]
+	if !ok {
+		return "", uuid.UUID{}, fmt.Errorf("No guid_root cmdline parameter")
+	}
+	for _, d := range devs {
+		fi, err := os.Stat(d)
+		if fi == nil || err != nil {
+			log.Print(err)
+			continue
+		}
+		if fi.Mode()&os.ModeType != os.ModeDevice {
+			log.Printf("%v is not a device", d)
+			continue
+		}
+		f, err := os.Open(d)
+		if err != nil {
+			log.Print(err)
+			continue
+		}
+		g, _, err := gpt.New(f)
+		f.Close()
+		if err != nil {
+			log.Print(err)
+			continue
+		}
+		// install media is always KERN-A, second partition.
+		if g.Parts[1].UniqueGUID.String() == rg {
+			log.Printf("%v: GUID %s matches for partition 2\n", d, rg)
+			return fmt.Sprintf("%s", d), g.Parts[1].UniqueGUID, nil
+		}
+	}
+	return "", uuid.UUID{}, fmt.Errorf("A device with that GUID was not found")
+}
+
+func main() {
+	flag.Parse()
+	if len(flag.Args()) < 1 {
+		flag.Usage()
+		log.Fatalf("install [options] dest-device")
+	}
+
+	parseCmdline()
+
+	d, u, err := findKernDev("/dev/sda", "/dev/sdb")
+	if err != nil {
+		log.Fatal(err)
+	}
+	// For now we always go with the B side. This means we need to change the
+	// GUID in the GPT for partition 4.
+	log.Printf("Install Media is on %s", d)
+
+	// We're going to be a bit paranoid in the order in which we do things.
+	// open the device and read the GPT.
+	// open both output partitions writeable to make sure that works.
+	// open input partitions.
+	// write the unchanged GPT back to the target as a test.
+	// write KERN-B
+	// write ROOT-B
+	// write the changed GPT with the new KERN-B UniqueGUID
+	// With luck, if there's a problem, we hit it long before
+	// we doing anything irreversible.
+
+	dest := flag.Args()[0]
+	destDev, err := os.OpenFile(dest, os.O_RDWR, 0)
+	if err != nil {
+		log.Fatal(err)
+	}
+	g, _, err := gpt.New(destDev)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// What a hack. mmc doesn't follow the venerable
+	// naming conventions.
+	var hack string
+	if strings.HasPrefix(dest, "/dev/mmc") {
+		log.Printf("It's mmc, add hack")
+		hack = "p"
+	}
+	log.Printf("Installing on %s%s{4,5}", dest, hack)
+	destKern, err := os.OpenFile(dest+hack+"4", os.O_RDWR, 0)
+	if err != nil {
+		log.Fatal(err)
+	}
+	destRoot, err := os.OpenFile(dest+hack+"5", os.O_RDWR, 0)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	kern, err := os.Open(d + "2")
+	if err != nil {
+		log.Fatal(err)
+	}
+	root, err := os.Open(d + "3")
+	if err != nil {
+		log.Fatal(err)
+	}
+	// Point of no return. Fix the GUID on the device.
+	// Write the old GPT back first to see if writes even work.
+	if err := gpt.Write(destDev, g); err != nil {
+		log.Fatal(err)
+	}
+	if _, err := io.Copy(destKern, kern); err != nil {
+		log.Fatal(err)
+	}
+	if _, err := io.Copy(destRoot, root); err != nil {
+		log.Fatal(err)
+	}
+	g.Parts[3].UniqueGUID = u
+	if err := gpt.Write(destDev, g); err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("All done. restart in chromeos and cgpt -i 4 -P 2 -S 0 -T 1 %v", dest)
```

Comment:
In the description this command has an `add`, is that required here as well?

Reply:
yes, you're right, nice catch, fixed!